### DISCUSSION TITLE
Default sort for box ?

### DISCUSSION THREAD

**Discussion by abhay-s-rawat** [isar/isar#1256] [Help] [answered]
What will be the default sort order for auto incremented ids ,like hive is sorted lexicographically for string keys.

And what will be the sort order(default) if I use belowfunction to convert String to int.
```
int fastHash(String string) {
    var hash = 0xcbf29ce484222325;
    var i = 0;
    while (i < string.length) {
      final codeUnit = string.codeUnitAt(i++);
      hash ^= codeUnit >> 8;
      hash *= 0x100000001b3;
      hash ^= codeUnit & 0xFF;
      hash *= 0x100000001b3;
    }
    return hash;
  }
```

**Comment by simc** [comment]
It's random. Use sortBy() to change the sort order

**Reply by abhay-s-rawat** [reply]
random for autoincrement ids or for fastHash function ?

**Reply by abhay-s-rawat** [reply]
That means it will read everything and then sort always ? Does hive do same (first reads and sorts by default in lexicographic order) ?

**Reply by simc** [reply]
Default sort order is by id but if you use hashes for ids they are effectively random.

**Reply by abhay-s-rawat** [reply]
is there a way to get alll records without filter or where with sort applied ?

**Reply by simc** [reply]
Like this: `yourCollection.where().sortByYourProperty().findAll()`

**Reply by abhay-s-rawat** [reply]
Thanks a lot bro, for all the effort you put for maintaining and developing hive and isar.
Seriously isar is super fast.
```
100000 records
write
Hive: 3465ms
ISAR: 2319ms

Read
Hive: 150/132 ms
ISAR: 62/50 ms

DB size (negligible)
Hive: 4.1 mb
Isar 5.2 mb
```
```
for 1 million records
write
Hive: 41000ms
ISAR: 22353ms

Read
Hive: 4567 ms
ISAR: 254 ms

DB size (negligible)
Hive: 41.9 mb
Isar: 47.2 mb
```
It is super fast super super fast. I am eagerly waiting for the encryption part to come up on isar, currently my project was 80% complete but now I have to change database and it will be a lot of work coming up hahaha. But it will be worth it absolute beast package.
Thanks a lot for your responses and time. Have a good day.

**Reply by abhay-s-rawat** [reply]
One last thing I would like to ask if I use a lexicographic sorted string as a key, will that fasthash function provide me sorted integers ?

**Reply by simc** [reply, marked as answer]
No, it's just a hash function. To my knowledge there exists no hash function that produces sorted keys. You can always use auto increment ids however and `@Index` your string key

**Reply by abhay-s-rawat** [reply]
One thing I noticed was If I make a field indexed rather than touching autogenerated id, I got speed improvements . This is may be due to the bigger fasthash integers generated.

So to get a single value from db. I did it with 2 approaches
1; fasthash a stiring id
2: id autogenerated and string field unique and indexed.

2nd approach is faster.

But what is the recommended approach to get only 1 value of id from database.
By approach 1 we can do  :  yourCollection.get(id)
By approach 2 I am doing : yourCollection.where().customIdEqualTo(45656).findAll() // is this query ok ? customId is indexed and unique

Can you please tell If I am using approach 2 then my selection for single record is ok .

**Reply by simc** [reply]
Both are fine. Isar 4 will introduce String ids

**Reply by abhay-s-rawat** [reply]
super cool. few more weeks for v4 ?

**Reply by abhay-s-rawat** [reply]
1 more thing does isar support extending collections like hive did with extends HiveObject keyword (.save() functiionality).

**Reply by abhay-s-rawat** [reply]
And 1 more thing to ask are you planning to implement map as supported data type in v4 ?
